File: sdk/packages/go/transport_ipc.go
package whodb

import (
	"bytes"
	"context"
	"encoding/json"
	"fmt"
	"net"
	"net/http"
	"os"
	"strings"
	"sync"
)

// IpcConfig configures the functions-runtime IPC transport. Zero values read
// the runtime's WHODB_IPC_ADDRESS / WHODB_JOB_ID / WHODB_IPC_TOKEN env vars.
type IpcConfig struct {
	Address string
	JobID   string
	Token   string
}

// IpcTransport runs the same facades inside the WhoDB Functions runtime over
// the in-container IPC server (unix socket in Docker, TCP in K8s). Only the
// ontology operations are available; others return ErrTransportCapability.
// GraphQL entity IDs resolve to apiNames via a cached /entities call.
type IpcTransport struct {
	address string
	jobID   string
	token   string
	client  *http.Client

	mu       sync.Mutex
	entities []map[string]any
}

// NewIpcTransport creates the IPC transport. WhoDB clients construct it
// automatically inside the Functions runtime; explicit use is for tests.
func NewIpcTransport(config IpcConfig) *IpcTransport {
	transport := &IpcTransport{
		address: firstNonEmpty(config.Address, os.Getenv("WHODB_IPC_ADDRESS")),
		jobID:   firstNonEmpty(config.JobID, os.Getenv("WHODB_JOB_ID")),
		token:   firstNonEmpty(config.Token, os.Getenv("WHODB_IPC_TOKEN")),
	}
	if strings.HasPrefix(transport.address, "/") {
		socketPath := transport.address
		transport.client = &http.Client{Transport: &http.Transport{
			DialContext: func(ctx context.Context, _, _ string) (net.Conn, error) {
				var dialer net.Dialer
				return dialer.DialContext(ctx, "unix", socketPath)
			},
		}}
	} else {
		transport.client = http.DefaultClient
	}
	return transport
}

func (t *IpcTransport) baseURL() string {
	if strings.HasPrefix(t.address, "/") {
		return "http://whodb-ipc"
	}
	return "http://" + t.address
}

func (t *IpcTransport) post(ctx context.Context, path string, body any) (any, error) {
	encoded, err := json.Marshal(body)
	if err != nil {
		return nil, err
	}
	request, err := http.NewRequestWithContext(ctx, http.MethodPost, t.baseURL()+path, bytes.NewReader(encoded))
	if err != nil {
		return nil, err
	}
	request.Header.Set("Content-Type", "application/json")
	request.Header.Set("X-Job-ID", t.jobID)
	request.Header.Set("Authorization", t.token)
	response, err := t.client.Do(request)
	if err != nil {
		return nil, err
	}
	defer response.Body.Close()
	if response.StatusCode >= 400 {
		return nil, &PlatformError{Message: fmt.Sprintf("IPC request %s failed with HTTP %d", path, response.StatusCode), Code: fmt.Sprintf("IPC_%d", response.StatusCode)}
	}
	var result any
	if err := json.NewDecoder(response.Body).Decode(&result); err != nil {
		return nil, &PlatformError{Message: "IPC request " + path + " returned invalid JSON", Code: "IPC_INVALID_JSON"}
	}
	return result, nil
}

func (t *IpcTransport) entityList(ctx context.Context) ([]map[string]any, error) {
	t.mu.Lock()
	defer t.mu.Unlock()
	if t.entities != nil {
		return t.entities, nil
	}
	raw, err := t.post(ctx, "/entities", map[string]any{})
	if err != nil {
		return nil, err
	}
	entries, _ := raw.([]any)
	entities := make([]map[string]any, 0, len(entries))
	for _, entry := range entries {
		entity, _ := entry.(map[string]any)
		entities = append(entities, entity)
	}
	t.entities = entities
	return entities, nil
}

func (t *IpcTransport) entityField(ctx context.Context, entityID, field string) (string, error) {
	entities, err := t.entityList(ctx)
	if err != nil {
		return "", err
	}
	for _, entity := range entities {
		if entity["id"] == entityID {
			value, _ := entity[field].(string)
			return value, nil
		}
	}
	return "", fmt.Errorf("%w: ontology entity %s not found in this function's scope", ErrNotFound, entityID)
}

func recordInputsToData(values any) map[string]any {
	data := map[string]any{}
	entries, _ := values.([]any)
	for _, entry := range entries {
		record, _ := entry.(map[string]any)
		key, _ := record["Key"].(string)
		data[key] = record["Value"]
	}
	// Also accept the typed shape the facades build directly.
	if typed, ok := values.([]map[string]string); ok {
		for _, record := range typed {
			data[record["Key"]] = record["Value"]
		}
	}
	return data
}

// Execute maps supported operations onto IPC endpoints, reshaping results
// into the GraphQL data shape the generated core expects.
func (t *IpcTransport) Execute(ctx context.Context, operation, _ string, variables map[string]any) (map[string]any, error) {
	result, err := t.dispatch(ctx, operation, variables)
	if err != nil {
		return nil, err
	}
	return map[string]any{operation: result}, nil
}

func (t *IpcTransport) dispatch(ctx context.Context, operation string, variables map[string]any) (any, error) {
	switch operation {
	case "OntologyEntities":
		entities, err := t.entityList(ctx)
		if err != nil {
			return nil, err
		}
		generic := make([]any, 0, len(entities))
		for _, entity := range entities {
			generic = append(generic, entity)
		}
		return generic, nil
	case "OntologyQuery":
		input, _ := variables["input"].(map[string]any)
		body := map[string]any{}
		for key, value := range input {
			if value == nil {
				continue
			}
			body[key] = value
		}
		if whereJSON, ok := body["whereJson"].(string); ok {
			delete(body, "whereJson")
			var where map[string]any
			if err := json.Unmarshal([]byte(whereJSON), &where); err == nil {
				body["where"] = where
			}
		}
		return t.post(ctx, "/query", body)
	case "OntologyDescribe":
		input, _ := variables["input"].(map[string]any)
		return t.post(ctx, "/describe", input)
	case "OntologyAddRow":
		entity, err := t.entityField(ctx, fmt.Sprint(variables["entityId"]), "apiName")
		if err != nil {
			return nil, err
		}
		if _, err := t.post(ctx, "/create", map[string]any{"entity": entity, "data": recordInputsToData(variables["values"])}); err != nil {
			return nil, err
		}
		return map[string]any{"Status": true}, nil
	case "OntologyAddRows":
		entity, err := t.entityField(ctx, fmt.Sprint(variables["entityId"]), "apiName")
		if err != nil {
			return nil, err
		}
		rows := []map[string]any{}
		entries, _ := variables["rows"].([]map[string]any)
		for _, entry := range entries {
			rows = append(rows, recordInputsToData(entry["values"]))
		}
		ids, err := t.post(ctx, "/create_many", map[string]any{"entity": entity, "rows": rows, "idempotencyKey": variables["idempotencyKey"]})
		if err != nil {
			return nil, err
		}
		idList, _ := ids.([]any)
		return map[string]any{"inserted": len(idList), "ids": idList}, nil
	case "OntologyUpdateRow":
		entityID := fmt.Sprint(variables["entityId"])
		entity, err := t.entityField(ctx, entityID, "apiName")
		if err != nil {
			return nil, err
		}
		pkKey, _ := t.entityField(ctx, entityID, "primaryKey")
		data := recordInputsToData(variables["values"])
		pk := ""
		if pkKey != "" {
			pk = fmt.Sprint(data[pkKey])
			delete(data, pkKey)
		}
		if _, err := t.post(ctx, "/update", map[string]any{"entity": entity, "pk": pk, "data": data}); err != nil {
			return nil, err
		}
		return map[string]any{"Status": true}, nil
	case "OntologyDeleteRow":
		entityID := fmt.Sprint(variables["entityId"])
		entity, err := t.entityField(ctx, entityID, "apiName")
		if err != nil {
			return nil, err
		}
		pkKey, _ := t.entityField(ctx, entityID, "primaryKey")
		data := recordInputsToData(variables["values"])
		pk := fmt.Sprint(data[pkKey])
		if _, err := t.post(ctx, "/delete", map[string]any{"entity": entity, "pk": pk}); err != nil {
			return nil, err
		}
		return map[string]any{"Status": true}, nil
	case "OntologyFollowLink":
		entity, err := t.entityField(ctx, fmt.Sprint(variables["entityId"]), "apiName")
		if err != nil {
			return nil, err
		}
		return t.post(ctx, "/follow_link", map[string]any{
			"entity":   entity,
			"pk":       variables["pk"],
			"link":     variables["linkApiName"],
			"pageSize": variables["pageSize"],
			"offset":   variables["pageOffset"],
		})
	default:
		return nil, fmt.Errorf("%w: %s is not available inside the function runtime in v1", ErrTransportCapability, operation)
	}
}

File: sdk/packages/rust/src/transport_ipc.rs
//! IPC transport: runs the same facades inside the WhoDB Functions runtime.
//!
//! Only the ontology operations are available; others return
//! Error::TransportCapability. GraphQL entity IDs resolve to apiNames via a
//! cached /entities call. Docker exposes the IPC server on a unix socket —
//! not supported by this transport yet (ureq is TCP-only); the K8s runtime's
//! TCP address works. Rust functions in the Docker runtime keep the legacy
//! helpers until unix-socket support lands here.

use serde_json::{json, Map, Value};
use std::sync::Mutex;

use crate::error::{Error, Result};
use crate::transport::Transport;

/// Functions-runtime IPC transport (TCP).
pub struct IpcTransport {
    address: String,
    job_id: String,
    token: String,
    agent: ureq::Agent,
    entities: Mutex<Option<Vec<Value>>>,
}

impl IpcTransport {
    /// Builds the transport from the runtime's WHODB_IPC_* env vars.
    pub fn from_env() -> Self {
        IpcTransport {
            address: std::env::var("WHODB_IPC_ADDRESS").unwrap_or_default(),
            job_id: std::env::var("WHODB_JOB_ID").unwrap_or_default(),
            token: std::env::var("WHODB_IPC_TOKEN").unwrap_or_default(),
            agent: ureq::Agent::new_with_defaults(),
            entities: Mutex::new(None),
        }
    }

    fn post(&self, path: &str, body: &Value) -> Result<Value> {
        if self.address.starts_with('/') {
            return Err(Error::TransportCapability(
                "unix-socket IPC is not supported by the Rust SDK yet — use the TCP runtime"
                    .to_string(),
            ));
        }
        let url = format!("http://{}{}", self.address, path);
        let mut response = self
            .agent
            .post(&url)
            .header("Content-Type", "application/json")
            .header("X-Job-ID", &self.job_id)
            .header("Authorization", &self.token)
            .send_json(body)
            .map_err(|error| match error {
                ureq::Error::StatusCode(code) => Error::Platform {
                    message: format!("IPC request {path} failed with HTTP {code}"),
                    code: format!("IPC_{code}"),
                },
                other => Error::Transport(other.to_string()),
            })?;
        response
            .body_mut()
            .read_json::<Value>()
            .map_err(|_| Error::Platform {
                message: format!("IPC request {path} returned invalid JSON"),
                code: "IPC_INVALID_JSON".to_string(),
            })
    }

    fn entity_list(&self) -> Result<Vec<Value>> {
        let mut cache = self.entities.lock().expect("entities lock poisoned");
        if let Some(entities) = cache.as_ref() {
            return Ok(entities.clone());
        }
        let raw = self.post("/entities", &json!({}))?;
        let entities = raw.as_array().cloned().unwrap_or_default();
        *cache = Some(entities.clone());
        Ok(entities)
    }

    fn entity_field(&self, entity_id: &str, field: &str) -> Result<String> {
        for entity in self.entity_list()? {
            if entity.get("id").and_then(Value::as_str) == Some(entity_id) {
                return Ok(entity
                    .get(field)
                    .and_then(Value::as_str)
                    .unwrap_or_default()
                    .to_string());
            }
        }
        Err(Error::NotFound(format!(
            "ontology entity {entity_id} not found in this function's scope"
        )))
    }
}

fn record_inputs_to_data(values: Option<&Value>) -> Map<String, Value> {
    let mut data = Map::new();
    if let Some(entries) = values.and_then(Value::as_array) {
        for entry in entries {
            if let Some(key) = entry.get("Key").and_then(Value::as_str) {
                data.insert(
                    key.to_string(),
                    entry.get("Value").cloned().unwrap_or(Value::Null),
                );
            }
        }
    }
    data
}

impl Transport for IpcTransport {
    fn execute(
        &self,
        operation: &str,
        _document: &str,
        variables: Map<String, Value>,
    ) -> Result<Value> {
        let payload = self.dispatch(operation, &variables)?;
        Ok(json!({ operation: payload }))
    }
}

impl IpcTransport {
    fn dispatch(&self, operation: &str, variables: &Map<String, Value>) -> Result<Value> {
        let entity_of = |key: &str| -> Result<String> {
            let entity_id = variables
                .get(key)
                .and_then(Value::as_str)
                .unwrap_or_default();
            self.entity_field(entity_id, "apiName")
        };
        match operation {
            "OntologyEntities" => Ok(Value::Array(self.entity_list()?)),
            "OntologyQuery" => {
                let input = variables
                    .get("input")
                    .and_then(Value::as_object)
                    .cloned()
                    .unwrap_or_default();
                let mut body = Map::new();
                for (key, value) in input {
                    if value.is_null() {
                        continue;
                    }
                    if key == "whereJson" {
                        if let Some(where_json) = value.as_str() {
                            if let Ok(parsed) = serde_json::from_str::<Value>(where_json) {
                                body.insert("where".to_string(), parsed);
                            }
                        }
                        continue;
                    }
                    body.insert(key, value);
                }
                self.post("/query", &Value::Object(body))
            }
            "OntologyDescribe" => {
                let input = variables.get("input").cloned().unwrap_or_else(|| json!({}));
                self.post("/describe", &input)
            }
            "OntologyAddRow" => {
                let entity = entity_of("entityId")?;
                self.post(
                    "/create",
                    &json!({"entity": entity, "data": record_inputs_to_data(variables.get("values"))}),
                )?;
                Ok(json!({"Status": true}))
            }
            "OntologyAddRows" => {
                let entity = entity_of("entityId")?;
                let rows: Vec<Value> = variables
                    .get("rows")
                    .and_then(Value::as_array)
                    .map(|entries| {
                        entries
                            .iter()
                            .map(|entry| Value::Object(record_inputs_to_data(entry.get("values"))))
                            .collect()
                    })
                    .unwrap_or_default();
                let ids = self.post(
                    "/create_many",
                    &json!({"entity": entity, "rows": rows, "idempotencyKey": variables.get("idempotencyKey")}),
                )?;
                let count = ids.as_array().map(Vec::len).unwrap_or(0);
                Ok(json!({"inserted": count, "ids": ids}))
            }
            "OntologyUpdateRow" => {
                let entity_id = variables
                    .get("entityId")
                    .and_then(Value::as_str)
                    .unwrap_or_default();
                let entity = self.entity_field(entity_id, "apiName")?;
                let pk_key = self
                    .entity_field(entity_id, "primaryKey")
                    .unwrap_or_default();
                let mut data = record_inputs_to_data(variables.get("values"));
                let pk = data
                    .remove(&pk_key)
                    .and_then(|value| value.as_str().map(str::to_string))
                    .unwrap_or_default();
                self.post(
                    "/update",
                    &json!({"entity": entity, "pk": pk, "data": data}),
                )?;
                Ok(json!({"Status": true}))
            }
            "OntologyDeleteRow" => {
                let entity_id = variables
                    .get("entityId")
                    .and_then(Value::as_str)
                    .unwrap_or_default();
                let entity = self.entity_field(entity_id, "apiName")?;
                let pk_key = self
                    .entity_field(entity_id, "primaryKey")
                    .unwrap_or_default();
                let data = record_inputs_to_data(variables.get("values"));
                let pk = data
                    .get(&pk_key)
                    .and_then(Value::as_str)
                    .unwrap_or_default();
                self.post("/delete", &json!({"entity": entity, "pk": pk}))?;
                Ok(json!({"Status": true}))
            }
            "OntologyFollowLink" => {
                let entity = entity_of("entityId")?;
                self.post(
                    "/follow_link",
                    &json!({
                        "entity": entity,
                        "pk": variables.get("pk"),
                        "link": variables.get("linkApiName"),
                        "pageSize": variables.get("pageSize"),
                        "offset": variables.get("pageOffset"),
                    }),
                )
            }
            other => Err(Error::TransportCapability(format!(
                "{other} is not available inside the function runtime in v1"
            ))),
        }
    }
}
